Spigot-Server-Patches/0419-Ticket-level-per-player-view-distance-API.patch

```diff
@@ -0,0 +1,745 @@
+diff --git a/src/main/java/net/minecraft/server/PlayerChunkMap.java b/src/main/java/net/minecraft/server/PlayerChunkMap.java
+index 59e74900..5ea8ecc3 100644
+--- a/src/main/java/net/minecraft/server/PlayerChunkMap.java
++++ b/src/main/java/net/minecraft/server/PlayerChunkMap.java
+@@ -935,30 +935,33 @@ public class PlayerChunkMap extends IChunkLoader implements PlayerChunk.d {
+ 
+     protected void setViewDistance(int i) {
+         int j = MathHelper.clamp(i + 1, 3, 33);
++        //Paper start ticket level pp-view-dist
++        this.chunkDistanceManager.a(j);
++        this.viewDistance = j;
+ 
+-        if (j != this.viewDistance) {
+-            int k = this.viewDistance;
+-
+-            this.viewDistance = j;
+-            this.chunkDistanceManager.a(this.viewDistance);
+-            ObjectIterator objectiterator = this.updatingChunks.values().iterator();
++        if(this.world.players != null && !this.world.players.isEmpty()) {
++            for (EntityPlayer player : this.world.players) {
++                player.setViewDistance(j);
++            }
++        }
++        //Paper end
++    }
+ 
+-            while (objectiterator.hasNext()) {
+-                PlayerChunk playerchunk = (PlayerChunk) objectiterator.next();
+-                ChunkCoordIntPair chunkcoordintpair = playerchunk.i();
++    //Paper start ticket level pp-view-dist
++    public void updateTrackViewDistance(EntityPlayer player, int i) {
++        for (int x = player.chunkX - i; x <= player.chunkX + i; ++x) {
++            for (int z = player.chunkZ - i; z <= player.chunkZ + i; ++z) {
++                ChunkCoordIntPair chunkcoordintpair = new ChunkCoordIntPair(x, z);
+                 Packet<?>[] apacket = new Packet[2];
++                int i1 = b(chunkcoordintpair, player, true);
++                boolean flag = i1 <= player.lastTrackViewDistance;
++                boolean flag1 = i1 <= i;
+ 
+-                this.a(chunkcoordintpair, false).forEach((entityplayer) -> {
+-                    int l = b(chunkcoordintpair, entityplayer, true);
+-                    boolean flag = l <= k;
+-                    boolean flag1 = l <= this.viewDistance;
+-
+-                    this.sendChunk(entityplayer, chunkcoordintpair, apacket, flag, flag1);
+-                });
++                this.sendChunk(player, chunkcoordintpair, apacket, flag, flag1);
+             }
+         }
+-
+     }
++    //Paper end
+ 
+     protected void sendChunk(EntityPlayer entityplayer, ChunkCoordIntPair chunkcoordintpair, Packet<?>[] apacket, boolean flag, boolean flag1) {
+         if (entityplayer.world == this.world) {
+@@ -973,6 +976,8 @@ public class PlayerChunkMap extends IChunkLoader implements PlayerChunk.d {
+                     }
+ 
+                     PacketDebug.a(this.world, chunkcoordintpair);
++                } else {
++                    entityplayer.a(chunkcoordintpair); //Paper ticket level pp-view-dist
```

Comment:
I'm not sure why you send unloads if there's no chunk holder

Reply:
if tracking distance > ticket distance and we track chunks with tickets loaded by other players, and that other player moves out of range of our tracking distance, we need to send an unload

Reply:
this is not how you implement that

Reply:
it would be great if you could suggest a solution instead

Reply:
maybe you can provide an explanation as to why this is incorrect instead of just pointing it out without following up

Reply:
I already said why it's not correct
`I'm not sure why you send unloads if there's no chunk holder`

Reply:
mojang always unloads anyways if it's out of range without a chunk holder, so adding this just follows suit

Reply:
if you have another way to implement this please suggest it

Reply:
we can just remove it though as it wouldn't cause an issue as chunks without chunk holders don't update

Reply:
removed in latest commit

Reply:
mojang only sends unloads once chunks go out of view distance range

I'm not going to write this PR for you
 - getVisibleChunk has nothing to do with player visibility of the chunk, check the src of how it's used and updated
 - I've already stated previously exactly why you will want to unload chunks outside the ticking range that are not at entity-ticking level, I'm not going to repeat that
 - There are ways to solve this that don't involve unloading the chunk, but those are even more hard to maintain

chunks without holders are unloaded or pending unload

Reply:
asking you for suggestions when you solely point out issues does not mean writing out this PR for me, as I've written in every commit so far, and yes I know getvisiblechunk has nothing to do with player visibility of the chunk... checking the rest now.

Reply:
chunks outside the ticking range that are not at entity-ticking level are already set to unload? just tested it... not sure what you are basing this off here, post some information/code maybe?

Reply:
one core aspect of the 1.14 chunk system is that only chunks at ticket level >= 45 are queued to be unloaded

Reply:
this method of reviewing is very ineffective.

most conversations go like this.
> there is a problem here described in a way which requires obscure levels of terminology

confusion? disagreement? misunderstanding? ok maybe recognised problem, will attempt to solve?
attempt solution #1
> this is wrong

attempt solution #2 
> this is still wrong, a reworded explanation with imperatives, yet no hard raw description/pointers of code, instead still just obscure derived levels of terminology

attempt solution #3
> this is still wrong

instead of this which would be much more useful and efficient

> there is a problem here, here is a suggestion/nudge as to how you may fix it with pointers to hard code and information

ok, attempt solution #1 or maybe 2
> solved

that way we can move at a much more faster pace and have this approved much quicker

Reply:
instead of this which would be much more useful and efficient

> there is a problem here, here is a suggestion/nudge as to how you may fix it with pointers to hard code and information

ok, attempt solution #1 or maybe 2
> solved

that way we can move at a much more faster pace and have this approved much quicker

Reply:
I just read your last comment, will have a look at it

Reply:
When making a PR like this that touches many aspects of the chunk system, you should have a basic understanding of the chunk system - and I've had to explain some. I'm not going further by giving solutions.

Reply:
these aren't solutions. these are suggestions.

Reply:
in a PR like this our terminology may not align. it is far more superior to point to code instead

Reply:
the moment you point to code I won't need to dig through and guess the terminology, instead I'll analyse it straight away and form a solution. this is much more efficient

Reply:
Check PlayerChunk#tickingFuture and the like

Reply:
Leaf u suck!

Reply:
> Leaf u suck!

no... what he says is substantial and useful. we just needed some clarification

Reply:
>Leaf u suck!

Be a better person and give better suggestions than Spottedleaf does then.

Reply:
33 33 33 33 33 33 33 34 35 35 35 35 35 35
33 32 32 32 32 32 33 34 34 34 34 34 34 34
33 32 31 31 31 32 33 33 33 33 33 33 33 34
33 32 31 p1 31 32 33 32 32 32 32 32 33 34 
33 32 31 31 31 32 33 32 31 31 31 32 33 34
33 32 32 32 32 32 33 32 31 p2 31 32 33 34
33 33 33 33 33 33 33 32 31 31 31 32 33 34
34 34 34 34 34 34 33 32 32 32 32 32 33 34
35 35 35 35 35 34 33 33 33 33 33 33 33 34
36 36 36 36 35 34 34 34 34 34 34 34 34 34 .... etc


so the issue here is if p2 has a tracking distance of 20 he will be sent and checked against all non ticking & entity ticking chunks such as 34, 35, 36.... to 44 (these all still in visiblechunks)

in a case like this then I assume we would want to emulate mojang's behavior of only sending and checking against all chunks with a ticket level of <= 32, as mojang adds tickets with <= vd - 2, we will need to adjust all our vd checks to check against this ticket level instead (entity trackers/spawning)

this should be the correct way to implement this.

Reply:
implemented